### PR TITLE
allowed value constraints

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

Adjust allowed value constraints and unit tests, include valid and invalid SSPS
### All Submissions:

- [ ] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [ ] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [ ] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                <enum value="other">Other</enum>
+            </allowed-values>
+
+            <allowed-values id="virtual" target="system-implementation/inventory-item/prop[@name='virtual']/@value" allow-other="no" level="ERROR">
+                <formal-name>Virtual</formal-name>
+                <description>Indicates if the asset is virtual.</description>
+                <enum value="yes">Yes</enum>
+                <enum value="no">No</enum>
+            </allowed-values>
+
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:142]: This already [exists](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_implementation-common_metaschema.xml#L137) in OSCAL and should be removed.
```suggestion
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                    <p>If the value is "no", the prop remarks must contain the reason why.</p>
+                </remarks>
+            </allowed-values>
+
+            <allowed-values id="public" target="system-implementation//prop[@name='public']/@value" allow-other="no" level="ERROR">
+                <formal-name>Public</formal-name>
+                <description>Indicates if the asset is exposed to the public Internet.</description>
+                <enum value="yes">Yes</enum>
+                <enum value="no">No</enum>
+            </allowed-values>
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:125]: This already [exists](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_implementation-common_metaschema.xml#L132) in OSCAL and should be removed.
```suggestion
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+            <allowed-values id="allows-authenticated-scan" target="system-implementation//prop[@name='allows-authenticated-scan']/@value" allow-other="no" level="ERROR">
+                <formal-name>Allows Authenticated Scan</formal-name>
+                <description>Indicates if the asset is capable of having an authenticated scan.</description>
+                <enum value="yes">Yes</enum>
+                <enum value="no">No</enum>
+                <remarks>
+                    <p>If the value is "no", the prop remarks must contain the reason why.</p>
+                </remarks>
+            </allowed-values>
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:118]: This [exists](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_implementation-common_metaschema.xml#L127) in OSCAL and should be removed.
```suggestion
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+            <allowed-values id="control-origination" target="control-implementation/implemented-requirement/prop[@name='control-origination'][@ns='https://fedramp.gov/ns/oscal']/@value" allow-other="no" level="ERROR">
+                <formal-name>Control Origination</formal-name>
+                <description>The point(s) from which the control satisfaction originates.</description>
+                <enum value="sp-corporate">Service Provider (Corporate)</enum>
+                <enum value="sp-system">Service Provider (System Specific)</enum>
+                <enum value="customer-configured">Configured by Customer</enum>
+                <enum value="customer-provided">Provided by Customer</enum>
+                <enum value="inherited">Inherited</enum>
+            </allowed-values>
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:88]: This [exists](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_ssp_metaschema.xml#L727) in OSCAL and should be removed.
```suggestion
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                <enum value="plan">Plan</enum>
+                <enum value="guidance">Guidance</enum>
+                <enum value="standard">Standard</enum>
+                <enum value="validation">Validation</enum>
+                <enum value="this-system">This System</enum>
+                <enum value="interconnection">Interconnection</enum>
+                <enum value="system">Leveraged System</enum>
+            </allowed-values>
+
+            <allowed-values id="control-implementation-status" target="control-implementation/implemented-requirement/prop[@name='implementation-status']/@value" allow-other="no" level="ERROR">
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:70]: This is located in the wrong place in the SSP model. It should be under [`by-component`](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_ssp_metaschema.xml#L852). The [current](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_implementation-common_metaschema.xml#L341) OSCAL constraint is `allow-other="yes"`, so this constraint is needed to close the values (i.e., `allow-other="no"`), so we should keep it, but target it differently.

```suggestion
            <allowed-values id="control-implementation-status" target="control-implementation/implemented-requirement/by-component/prop[@name='implementation-status']/@value" allow-other="no" level="ERROR">
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                <enum value="process">Process</enum>
+                <enum value="procedure">Procedure</enum>
+                <enum value="plan">Plan</enum>
+                <enum value="guidance">Guidance</enum>
+                <enum value="standard">Standard</enum>
+                <enum value="validation">Validation</enum>
+                <enum value="this-system">This System</enum>
+                <enum value="interconnection">Interconnection</enum>
+                <enum value="system">Leveraged System</enum>
+            </allowed-values>
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:68]: The [current](https://github.com/usnistgov/OSCAL/blob/4f02dac6f698efda387cc5f55bc99581eaf494b6/src/metaschema/oscal_implementation-common_metaschema.xml#L236) OSCAL constraint is allow-other="yes", so this constraint is needed to close the values (i.e., allow-other="no"), so we should keep it.

Our values are out of sync with the OSCAL values, so we need to confirm that the documentation is correct.
```suggestion

            <allowed-values id="component-type" target="system-implementation/component/@type" allow-other="no" level="ERROR">
                <formal-name>Component Type</formal-name>
                <description>Identifies the component type.</description>
                <enum value="interconnection">A connection to something outside this system.</enum>
                <enum value="software">Any software, operating system, or firmware.</enum>
                <enum value="hardware">A physical device.</enum>
                <enum value="service">A service that may provide APIs.</enum>
                <enum value="policy">An enforceable policy.</enum>
                <enum value="physical">A tangible asset used to provide physical protections or countermeasures.</enum>
                <enum value="process-procedure">A list of steps or actions to take to achieve some end result.</enum>
                <enum value="plan">An applicable plan.</enum>
                <enum value="guidance">Any guideline or recommendation.</enum>
                <enum value="standard">Any organizational or industry standard.</enum>
                <enum value="validation">An external assessment performed on some other component, that has been validated by a third-party.</enum>
                <enum value="this-system">The system as a whole.</enum>
                <enum value="system">An external system, which may be a leveraged system or the other side of an interconnection.</enum>
                <enum value="network">A physical or virtual network.</enum>
            </allowed-values>
```

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+<metaschema-meta-constraints xmlns="http://csrc.nist.gov/ns/oscal/metaschema/1.0">
+    <!-- ================== -->
+    <!-- FedRAMP Extensions -->
+    <!-- ================== -->
+
+    <context>
+        <metapath target="/system-security-plan"/>
+
+        <constraints>
+            <allowed-values id="attachment-type" target="back-matter/resource/prop[@name='type']/@value" allow-other="yes" level="ERROR">
```
Comment (david-waltermire) [inline src/validations/constraints/fedramp-external-allowed-values-DZ.xml:11]: This constraint is not contextualized correctly. This applies to a back-matter resource. Please move this to a different context.

File: src/validations/constraints/content/ssp-all-VALID.xml [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    </implemented-requirement>
+  </control-implementation>
+  
+  <back-matter>
+    <resource uuid="eeeeeeee-0000-4000-9000-00000000000e">
+      <title>Access Control Policy</title>
+      <description>
+        <p>Detailed access control policy document</p>
+      </description>
+      <prop name="type" value="policy" ns="https://fedramp.gov/ns/oscal"/>
```
Comment (david-waltermire) [inline src/validations/constraints/content/ssp-all-VALID.xml:188]: This is not an invalid case to test the resource constraint.

**Review by david-waltermire — CHANGES_REQUESTED** [review]
I provided some feedback on individual constraints.

Some are targeted improperly and their context or target need to be updated as indicated in my comments.  It also looks like the negative tests are not sufficient, since these were passing. We need to add negative test cases for these to confirm proper operation.

Others duplicate existing OSCAL constraints, which we don't need unless we are trying to close the value list for something that is `allow-other="yes"`. If the value set is already closed, we need to mark these as "drop" in the tracker and remove them as indicated.

**Comment by david-waltermire** [comment]
This looks like good progress. Overall, we need to check that every constraint has both a negative and positive test case. There we some in this set that are improperly targeted, which may mean the testing is not robust enough to find the problem.

File: src/validations/constraints/content/ssp-all-VALID.xml (hunk 8 above)
Comment (wandmagic) [reply src/validations/constraints/content/ssp-all-VALID.xml:193]: this is fixed

File: src/validations/constraints/fedramp-external-allowed-values-DZ.xml (hunk 7 above)
Comment (wandmagic) [reply src/validations/constraints/fedramp-external-allowed-values-DZ.xml]: what context does it belong to?
ssp.sch shows attachment type only applies to backmatter resources

**Review by wandmagic — COMMENTED** [review]
added more robust reporting and a new failure test option "mixed" for if a single failing test counts as a negative test case

**Review by wandmagic — COMMENTED** [review]
looks good to me now

**Review by wandmagic — COMMENTED** [review]
looks good

**Comment by wandmagic** [comment]
this is ready to be looked at again @david-waltermire

File: src/validations/constraints/content/profile-all-INVALID.xml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+      <with-id>ac-1</with-id>
+    </include-controls>
+  </import>
+  <modify>
+    <alter control-id="ac-1">
+      <add position="starting">
+        <part name="statement">
+          <prop name="response-point" ns="https://fedramp.gov/ns/oscal" value="123"/>
+          <prop name="response-point" ns="https://fedramp.gov/ns/oscal" value="1234"/>
+          <p>This is a test statement with a non-string response point.</p>
```
Comment (Rene2mt) [inline src/validations/constraints/content/profile-all-INVALID.xml:21]: We currently don't have any constraints around `response-point` value, however, a given (control) part shouldn't have more than one `response-point`.  
```suggestion
          <p>This is a test checking that profiles validation fails if more than one response point is specified for a given (control) part.</p>
```

File: src/validations/constraints/content/test-valid-response-point-use.xml
Comment (Rene2mt) [inline src/validations/constraints/content/test-valid-response-point-use.xml:1]: For consistency, recommend we use the following naming convention for validation constraint file names:
* *{model}-{assembly, field, or flag name}-{VALID or INVALID}.xml*.  Following this convention, this unit test content file would be renamed **profile-response-point-INVALID.xml**

Following the same pattern, instances where the content covers **all** data in an OSCAL document, the naming convention would be *{model}-all-{VALID or INVALID}.  For example **profile-all-INVALID.xml**.

File: src/validations/constraints/unit-tests/response-point-FAIL.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# driver for the response-point unit test
+test-case:
+  name: Test Valid Response Point Use
+  description: Test that the resolved profile contains valid response-points.
+  content: ../content/test-valid-response-point-use.xml # the content to run
```
Comment (Rene2mt) [inline src/validations/constraints/unit-tests/response-point-FAIL.yaml:5]: see other comment on naming convention for validation constraint content file names.  If the file is changes, please also update the reference here

```suggestion
content: ../content/profile-response-point-INVALID.xml # the content to run
```

**Review by Rene2mt — CHANGES_REQUESTED** [review]

File: src/validations/constraints/content/fedramp-tailoring-profile.xml [hunk 11]
```diff
@@ -5,11 +5,11 @@
     <published>2024-06-25T00:00:00Z</published>
     <last-modified>2024-06-25T00:00:00Z</last-modified>
     <version>2.0</version>
-    <prop ns="https://fedramp.gov/ns/oscal" name="fedramp-version" value="Required"/>
+    <prop ns="https://fedramp.gov/ns/oscal" name="fedramp-version" value="2.0.0"/>
```
[the remark this answers is not in the document]
Comment (wandmagic) [reply src/validations/constraints/content/fedramp-tailoring-profile.xml]: removed

**Review by david-waltermire — CHANGES_REQUESTED** [review]
It still looks like the mime types have not been addressed yet.

File: src/validations/module.mk [hunk 12]
```diff
@@ -8,11 +8,9 @@ REV5_TEMPLATES = ./dist/content/rev5/templates
...
 .PHONY: init-validations
 init-validations:
-	@echo "Installing OSCAL CLI..."
-	npm install oscal -g
-	npm install --only=dev
-	oscal use oscal-cli-2.0.2.rc1
-	oscal -V
+	@echo "Installing node modules..."
+	npm install
+	npx oscal@latest use oscal-cli-2.0.2.rc1
```
[the remark this answers is not in the document]
Comment (wandmagic) [reply src/validations/module.mk]: we no longer specify version here, we just use latest which should get automatically latest from maven

File: src/content/rev5/baselines/xml/FedRAMP_rev5_LI-SaaS-baseline_profile.xml [hunk 13]
```diff
@@ -2958,7 +2958,7 @@
         <resource uuid="051a77c1-b61d-4995-8275-dacfe688d510">
             <title>NIST Special Publication (SP) 800-53</title>
             <prop name="version" value="5.1.1"/>
-            <rlink media-type="application/oscal+xml" href="https://raw.githubusercontent.com/usnistgov/oscal-content/v1.2.0/nist.gov/SP800-53/rev5/xml/NIST_SP-800-53_rev5_catalog.xml"/>
+            <rlink media-type="application/xml" href="https://raw.githubusercontent.com/usnistgov/oscal-content/v1.2.0/nist.gov/SP800-53/rev5/xml/NIST_SP-800-53_rev5_catalog.xml"/>
```
[the remark this answers is not in the document]
Comment (wandmagic) [reply src/content/rev5/baselines/xml/FedRAMP_rev5_LI-SaaS-baseline_profile.xml]: it has been changed to oscal+xml in the latest version

File: .github/workflows/run-tests.yml [hunk 14]
```diff
@@ -16,16 +16,23 @@ on:
 jobs:
   # one job that runs tests
   run-tests:
-    runs-on: ubuntu-20.04
     strategy:
       fail-fast: false
-
+      matrix:
+        os: [ubuntu-20.04, windows-latest]
```
Comment (david-waltermire) [inline .github/workflows/run-tests.yml:22]: Can we pin to a windows version?

Comment (wandmagic) [reply .github/workflows/run-tests.yml:22]: pinned

**Comment by david-waltermire** [comment]
As follow-on work, #628 will be used to address OSCAL CLI versioning.

File: src/validations/module.mk (hunk 12 above)
[the remark this answers is not in the document]
Comment (wandmagic) [reply src/validations/module.mk]: See issue  #628 for tracking a better resolution to this issue in a future PR

File: src/validations/module.mk [hunk 15]
```diff
@@ -8,10 +8,9 @@ REV5_TEMPLATES = ./dist/content/rev5/templates
 # Preparation
 .PHONY: init-validations
 init-validations:
-	@echo "Installing OSCAL CLI..."
-	npm install oscal -g
-	npm install --only=dev
-	oscal -V
+	@echo "Installing node modules..."
+	npm install
+	$(OSCAL_CLI) use latest
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply src/validations/module.mk:13]: See #628

File: src/validations/module.mk [hunk 16]
```diff
@@ -1,5 +1,5 @@
 # Variables
-OSCAL_CLI = oscal
+OSCAL_CLI = npx oscal@latest
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply src/validations/module.mk:2]: See #628

File: src/content/module.mk [hunk 17]
```diff
@@ -8,7 +8,8 @@ YAML_DIR = $(DIST_DIR)/content/rev5/baselines/yaml
 
 .PHONY: init-content
 init-content:
-	@npm install oscal -g
+	@npm install
+	$(OSCAL_CLI) use latest
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply src/content/module.mk:12]: See #628

**Review by david-waltermire — CHANGES_REQUESTED** [review]

File: .github/workflows/run-tests.yml [hunk 18]
```diff
@@ -16,16 +16,23 @@ on:
...
+    runs-on: ${{ matrix.os }}
     # Checkout repository and its submodules
     steps:
       # Check-out the repository under $GITHUB_WORKSPACE
       - name: Checkout repository
         uses: actions/checkout@v2
 
+      - name: Set up Java
+        if: runner.os == 'Windows'
```
[the remark this answers is not in the document]
Comment (wandmagic) [reply .github/workflows/run-tests.yml:31]: it would take more time is all, it appears that ubuntu comes with java version suitable for oscal-cli

[the remark this answers is not in the document]
Comment (wandmagic) [reply .github/workflows/run-tests.yml]: added java to ubuntu and only took an extra 5 seconds

**Review by david-waltermire — APPROVED** [review]
